src/Nethermind/Nethermind.Consensus.AuRa/Validators/ReportingContractBasedValidator.Cache.cs

```diff
@@ -0,0 +1,52 @@
+﻿//  Copyright (c) 2018 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using Nethermind.Core;
+
+namespace Nethermind.Consensus.AuRa.Validators
+{
+    public partial class ReportingContractBasedValidator
+    {
+        public class Cache
+        {
+            internal LinkedList<PersistentReport> PersistentReports { get; } = new LinkedList<PersistentReport>();
+            
+            private long _lastReportedBlockNumber;
+            private readonly ConcurrentDictionary<(Address Validator, bool Malicious, long BlockNumber, object Cause), bool> _lastBlockReports = 
+                new ConcurrentDictionary<(Address Validator, bool Type, long BlockNumber, object cause), bool>(); 
+        
+            internal bool AlreadyReported(bool malicious, Address validator, in long blockNumber, object cause)
+            {
+                var lastReportedBlockNumber = Interlocked.Exchange(ref _lastReportedBlockNumber, blockNumber);
+                (Address Validator, bool Malicious, long BlockNumber, object Cause) key = (validator, malicious, blockNumber, cause);
+                if (lastReportedBlockNumber == blockNumber)
```

Comment:
you should Clear first  if block number is new and then for any case you should return TryAdd result